swarm/network/kademlia.go

```diff
@@ -536,71 +570,127 @@ func (k *Kademlia) Prune(c <-chan time.Time) {
 	np := pot.NewPot(nil, 0)
-	for _, id := range ids {
-		o := ToOverlayAddr(id.Bytes())
-		np, _, _ = pot.Add(np, o, pof)
+	for _, addr := range addrs {
+		np, _, _ = pot.Add(np, addr, pof)
 	}
-	nnmap := make(map[discover.NodeID][][]byte)
+	ppmap := make(map[discover.NodeID]*PeerPot)
 
-	for _, id := range ids {
-		pl := 0
+	for i, id := range ids {
+		pl := 256
+		prev := 256
+		var emptyBins []int
 		var nns [][]byte
-		np.EachNeighbour(id.Bytes(), pof, func(val pot.Val, po int) bool {
-			// a := val.(pot.BytesAddress).Address()
-			// nns = append(nns, a)
+		np.EachNeighbour(addrs[i], pof, func(val pot.Val, po int) bool {
 			a := val.([]byte)
-			nns = append(nns, a)
-			if len(nns) >= kadMinProxSize {
+			if po == 256 {
+				return true
+			}
+			if pl == 256 || pl == po {
+				nns = append(nns, a)
+			}
+			if pl == 256 && len(nns) >= kadMinProxSize {
 				pl = po
+				prev = po
 			}
-			return pl == 0 || pl == po
+			if prev < pl {
+				for j := prev; j > po; j-- {
+					emptyBins = append(emptyBins, j)
+				}
+			}
+			prev = po - 1
+			return true
 		})
-		nnmap[id] = nns
+		for j := prev; j >= 0; j-- {
+			emptyBins = append(emptyBins, j)
+		}
+		log.Trace(fmt.Sprintf("%x NNS: %s", addrs[i][:4], logNNS(nns)))
+		ppmap[id] = &PeerPot{nns, emptyBins}
 	}
-	return nnmap
+	return ppmap
+}
+
+func (k *Kademlia) full(emptyBins []int) (full bool) {
+	prev := 0
+	e := len(emptyBins)
+	k.conns.EachBin(k.base, pof, 0, func(po, _ int, _ func(func(val pot.Val, i int) bool) bool) bool {
+		for i := prev; e > 0 && i < po; i++ {
+			e--
+			if emptyBins[e] != i {
+				log.Trace(fmt.Sprintf("%08x po: %d, i: %d, e: %d, emptybins: %v", k.BaseAddr()[:4], po, i, e, logEmptyBins(emptyBins)))
+				if emptyBins[e] < i {
+					panic("incorrect peerpot")
+				}
+				return false
+			}
+		}
+		prev = po + 1
+		return true
+	})
+	return e == 0
 }
 
-// FirstEmptyBin returns the farthest proximity order (int) that has no peer records
-func (k *Kademlia) firstEmptyBin() (i int) {
-	i = -1
-	k.conns.EachBin(k.base, pof, 0, func(po, size int, f func(func(val pot.Val, i int) bool) bool) bool {
-		if po > i+1 {
-			i = po
+func (k *Kademlia) gotNearestNeighbours(peers [][]byte) bool {
+	pm := make(map[string]bool)
+
+	k.eachConn(nil, 255, func(p OverlayConn, po int, nn bool) bool {
+		if !nn {
 			return false
 		}
-		i = po
+		pk := fmt.Sprintf("%x", p.Address())
+		pm[pk] = true
 		return true
 	})
-	return i
+	log.Trace(fmt.Sprintf("%08x: NNSet: %d", k.BaseAddr()[:4], len(pm)))
+	for _, p := range peers {
+		pk := fmt.Sprintf("%x", p)
+		if !pm[pk] {
+			log.Trace(fmt.Sprintf("%08x: ExpNN: %s not found", k.BaseAddr()[:4], pk))
+			return false
+		}
+	}
+	return true
 }
 
-// Full returns a bool if the kademlia table is healthy and complete
-func (k *Kademlia) full() bool {
-	return k.firstEmptyBin() >= k.depth()
+// Healthy reports the health state of the kademlia connectivity
+func (k *Kademlia) Healthy(pp *PeerPot) bool {
```

Comment:
Here the Healthy method takes a struct labelled for testing only. Is this correct dependency?